File: server/relay_server.js
/**
 * Anger of Stick 5 - Ultra-Low-Latency 4-Player UDP Relay Server
 * Protocol: Binary UDP
 * Default Port: 7777
 */

const dgram = require('dgram');

const PORT = parseInt(process.env.PORT || '7777', 10);
const HOST = process.env.HOST || '0.0.0.0';
const HEARTBEAT_TIMEOUT_MS = 10000; // 10 seconds timeout

// Protocol constants
const MAGIC_BYTE = 0x53; // 'S'
const PROTO_VERSION = 0x01;

const PacketType = {
    JOIN_REQUEST: 0x01,
    JOIN_RESPONSE: 0x02,
    HEARTBEAT: 0x03,
    PLAYER_STATE: 0x04,
    ACTION_EVENT: 0x05,
    DAMAGE_EVENT: 0x06,
    SPAWN_PROXY: 0x07,
    DISCONNECT: 0x08,
    ROOM_LIST: 0x09
};

// Rooms storage: roomId -> Room object
// Room object: { id, createdAt, peers: Map<peerId, { address, port, lastSeen, slot }> }
const rooms = new Map();

// Client endpoint key -> { roomId, peerId }
const clientMap = new Map();

function getEndpointKey(address, port) {
    return `${address}:${port}`;
}

const server = dgram.createSocket('udp4');

server.on('error', (err) => {
    console.error(`[Server Error] ${err.stack}`);
    server.close();
});

server.on('message', (msg, rinfo) => {
    if (msg.length < 3) return;

    const magic = msg.readUInt8(0);
    const version = msg.readUInt8(1);
    const packetType = msg.readUInt8(2);

    if (magic !== MAGIC_BYTE || version !== PROTO_VERSION) {
        return; // Invalid packet
    }

    const endpointKey = getEndpointKey(rinfo.address, rinfo.port);

    switch (packetType) {
        case PacketType.JOIN_REQUEST: {
            handleJoinRequest(msg, rinfo, endpointKey);
            break;
        }
        case PacketType.HEARTBEAT: {
            handleHeartbeat(msg, rinfo, endpointKey);
            break;
        }
        case PacketType.PLAYER_STATE:
        case PacketType.ACTION_EVENT:
        case PacketType.DAMAGE_EVENT:
        case PacketType.SPAWN_PROXY: {
            handleRelayPacket(msg, rinfo, endpointKey, packetType);
            break;
        }
        case PacketType.DISCONNECT: {
            handleDisconnect(msg, rinfo, endpointKey);
            break;
        }
        default: {
            console.warn(`[Unknown Packet] Type 0x${packetType.toString(16)} from ${endpointKey}`);
            break;
        }
    }
});

function handleJoinRequest(msg, rinfo, endpointKey) {
    // Expected payload: magic(1) + ver(1) + type(1) + roomId(4) + preferredSlot(1)
    let roomId = 1;
    let preferredSlot = 0xFF; // 255 = any

    if (msg.length >= 7) {
        roomId = msg.readUInt32LE(3);
    }
    if (msg.length >= 8) {
        preferredSlot = msg.readUInt8(7);
    }

    let room = rooms.get(roomId);
    if (!room) {
        room = {
            id: roomId,
            createdAt: Date.now(),
            peers: new Map() // slot (0..3) -> client info
        };
        rooms.set(roomId, room);
        console.log(`[Room Created] Room ID: ${roomId}`);
    }

    // Check if client is already in room
    let assignedSlot = -1;
    for (const [slot, peer] of room.peers.entries()) {
        if (peer.endpointKey === endpointKey) {
            assignedSlot = slot;
            break;
        }
    }

    if (assignedSlot === -1) {
        // Find available slot (0 to 3)
        if (preferredSlot >= 0 && preferredSlot <= 3 && !room.peers.has(preferredSlot)) {
            assignedSlot = preferredSlot;
        } else {
            for (let s = 0; s < 4; s++) {
                if (!room.peers.has(s)) {
                    assignedSlot = s;
                    break;
                }
            }
        }
    }

    if (assignedSlot === -1) {
        // Room is full
        console.log(`[Join Rejected] Room ${roomId} full for ${endpointKey}`);
        sendJoinResponse(rinfo, 0x01 /* FULL */, 0xFF, roomId);
        return;
    }

    // Register peer
    const now = Date.now();
    const peerInfo = {
        slot: assignedSlot,
        address: rinfo.address,
        port: rinfo.port,
        endpointKey: endpointKey,
        lastSeen: now
    };
    room.peers.set(assignedSlot, peerInfo);
    clientMap.set(endpointKey, { roomId, peerId: assignedSlot });

    console.log(`[Peer Joined] Endpoint: ${endpointKey} assigned Slot: ${assignedSlot} in Room: ${roomId} (Total: ${room.peers.size}/4)`);
    sendJoinResponse(rinfo, 0x00 /* SUCCESS */, assignedSlot, roomId);

    // Notify all other peers in the room about the new peer
    const spawnNotice = Buffer.alloc(13);
    spawnNotice.writeUInt8(MAGIC_BYTE, 0);
    spawnNotice.writeUInt8(PROTO_VERSION, 1);
    spawnNotice.writeUInt8(PacketType.SPAWN_PROXY, 2);
    spawnNotice.writeUInt8(assignedSlot, 3);
    spawnNotice.writeUInt8(0 /* default hero */, 4);
    spawnNotice.writeFloatLE(0.0, 5); // x
    spawnNotice.writeFloatLE(0.0, 9); // y
    broadcastToRoom(room, spawnNotice, assignedSlot);
}

function sendJoinResponse(rinfo, status, assignedSlot, roomId) {
    // Response: magic(1) + ver(1) + type(1) + status(1) + peerId(1) + roomId(4)
    const resp = Buffer.alloc(9);
    resp.writeUInt8(MAGIC_BYTE, 0);
    resp.writeUInt8(PROTO_VERSION, 1);
    resp.writeUInt8(PacketType.JOIN_RESPONSE, 2);
    resp.writeUInt8(status, 3);
    resp.writeUInt8(assignedSlot, 4);
    resp.writeUInt32LE(roomId, 5);

    server.send(resp, 0, resp.length, rinfo.port, rinfo.address);
}

function handleHeartbeat(msg, rinfo, endpointKey) {
    const client = clientMap.get(endpointKey);
    if (!client) return;

    const room = rooms.get(client.roomId);
    if (!room) return;

    const peer = room.peers.get(client.peerId);
    if (peer) {
        peer.lastSeen = Date.now();
    }

    // Echo heartbeat back
    server.send(msg, 0, msg.length, rinfo.port, rinfo.address);
}

function handleRelayPacket(msg, rinfo, endpointKey, packetType) {
    const client = clientMap.get(endpointKey);
    if (!client) return;

    const room = rooms.get(client.roomId);
    if (!room) return;

    const peer = room.peers.get(client.peerId);
    if (peer) {
        peer.lastSeen = Date.now();
    }

    // Fan-out to all OTHER peers in the room
    broadcastToRoom(room, msg, client.peerId);
}

function handleDisconnect(msg, rinfo, endpointKey) {
    const client = clientMap.get(endpointKey);
    if (!client) return;

    removeClient(endpointKey, client.roomId, client.peerId);
}

function broadcastToRoom(room, buffer, senderSlot) {
    for (const [slot, peer] of room.peers.entries()) {
        if (slot !== senderSlot) {
            server.send(buffer, 0, buffer.length, peer.port, peer.address);
        }
    }
}

function removeClient(endpointKey, roomId, peerId) {
    clientMap.delete(endpointKey);
    const room = rooms.get(roomId);
    if (room) {
        room.peers.delete(peerId);
        console.log(`[Peer Left] Slot ${peerId} left Room ${roomId}. Remaining: ${room.peers.size}/4`);

        // Broadcast disconnect to remaining peers
        const discMsg = Buffer.alloc(4);
        discMsg.writeUInt8(MAGIC_BYTE, 0);
        discMsg.writeUInt8(PROTO_VERSION, 1);
        discMsg.writeUInt8(PacketType.DISCONNECT, 2);
        discMsg.writeUInt8(peerId, 3);
        broadcastToRoom(room, discMsg, peerId);

        if (room.peers.size === 0) {
            rooms.delete(roomId);
            console.log(`[Room Destroyed] Room ${roomId} empty and cleaned up.`);
        }
    }
}

// Timeout garbage collection for inactive peers
setInterval(() => {
    const now = Date.now();
    for (const [roomId, room] of rooms.entries()) {
        for (const [slot, peer] of room.peers.entries()) {
            if (now - peer.lastSeen > HEARTBEAT_TIMEOUT_MS) {
                console.log(`[Timeout Disconnect] Peer ${peer.endpointKey} (Slot ${slot}) timed out.`);
                removeClient(peer.endpointKey, roomId, slot);
            }
        }
    }
}, 3000);

server.on('listening', () => {
    const addr = server.address();
    console.log(`=======================================================`);
    console.log(`  Anger of Stick 5 UDP Relay Server Listening on ${addr.address}:${addr.port}`);
    console.log(`  Protocol Version: 0x01 | Magic: 0x53 ('S')`);
    console.log(`  Room Capacity: 4 Players per Room`);
    console.log(`=======================================================`);
});

server.bind(PORT, HOST);

File: server/relay_server.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <chrono>
#include <cstring>
#include <cstdint>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <unistd.h>
#include <fcntl.h>

#pragma pack(push, 1)
struct PacketHeader {
    uint8_t magic;    // 0x53 ('S')
    uint8_t version;  // 0x01
    uint8_t type;     // PacketType
};

struct JoinRequestPacket {
    PacketHeader header;
    uint32_t room_id;
    uint8_t preferred_slot;
};

struct JoinResponsePacket {
    PacketHeader header;
    uint8_t status; // 0 = SUCCESS, 1 = FULL
    uint8_t peer_id;
    uint32_t room_id;
};

struct PlayerStatePacket {
    PacketHeader header;
    uint8_t peer_id;
    float x;
    float y;
    uint8_t anim_id;
    uint8_t flip_x;
    int16_t hp;
    uint16_t flags;
    uint32_t seq;
};

struct ActionEventPacket {
    PacketHeader header;
    uint8_t peer_id;
    uint8_t action_id;
    float target_x;
    float target_y;
    int32_t extra;
};

struct DamageEventPacket {
    PacketHeader header;
    uint8_t attacker_id;
    uint8_t target_slot;
    uint8_t damage_type;
    int16_t damage_amount;
};
#pragma pack(pop)

enum PacketType : uint8_t {
    JOIN_REQUEST  = 0x01,
    JOIN_RESPONSE = 0x02,
    HEARTBEAT     = 0x03,
    PLAYER_STATE  = 0x04,
    ACTION_EVENT  = 0x05,
    DAMAGE_EVENT  = 0x06,
    SPAWN_PROXY   = 0x07,
    DISCONNECT    = 0x08
};

struct Peer {
    uint8_t slot;
    sockaddr_in addr;
    std::chrono::steady_clock::time_point last_seen;
    bool active;
};

struct Room {
    uint32_t id;
    Peer peers[4];
};

static uint64_t endpoint_hash(const sockaddr_in& addr) {
    return ((uint64_t)addr.sin_addr.s_addr << 32) | (uint64_t)addr.sin_port;
}

int main(int argc, char* argv[]) {
    int port = 7777;
    if (argc > 1) {
        port = std::atoi(argv[1]);
    }

    int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
    if (sockfd < 0) {
        std::cerr << "Failed to create socket\n";
        return 1;
    }

    // Set SO_REUSEADDR
    int opt = 1;
    setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt));

    sockaddr_in servaddr{};
    servaddr.sin_family = AF_INET;
    servaddr.sin_addr.s_addr = INADDR_ANY;
    servaddr.sin_port = htons(port);

    if (bind(sockfd, (const struct sockaddr*)&servaddr, sizeof(servaddr)) < 0) {
        std::cerr << "Bind failed on port " << port << "\n";
        close(sockfd);
        return 1;
    }

    std::cout << "=======================================================\n";
    std::cout << "  Anger of Stick 5 C++ Native UDP Relay Server (Port: " << port << ")\n";
    std::cout << "  Protocol: Binary UDP | Max 4 Players / Room\n";
    std::cout << "=======================================================\n";

    std::unordered_map<uint32_t, Room> rooms;
    std::unordered_map<uint64_t, std::pair<uint32_t, uint8_t>> client_to_room; // endpoint -> (roomId, slot)

    uint8_t buffer[2048];
    sockaddr_in cliaddr{};
    socklen_t len = sizeof(cliaddr);

    while (true) {
        ssize_t n = recvfrom(sockfd, buffer, sizeof(buffer), 0, (struct sockaddr*)&cliaddr, &len);
        if (n < (ssize_t)sizeof(PacketHeader)) continue;

        auto* hdr = (PacketHeader*)buffer;
        if (hdr->magic != 0x53 || hdr->version != 0x01) continue;

        uint64_t ep = endpoint_hash(cliaddr);
        auto now = std::chrono::steady_clock::now();

        if (hdr->type == JOIN_REQUEST && n >= (ssize_t)sizeof(JoinRequestPacket)) {
            auto* req = (JoinRequestPacket*)buffer;
            uint32_t room_id = req->room_id;
            uint8_t preferred = req->preferred_slot;

            auto& room = rooms[room_id];
            room.id = room_id;

            int assigned_slot = -1;
            // Check existing
            for (int s = 0; s < 4; ++s) {
                if (room.peers[s].active && endpoint_hash(room.peers[s].addr) == ep) {
                    assigned_slot = s;
                    break;
                }
            }

            if (assigned_slot == -1) {
                if (preferred < 4 && !room.peers[preferred].active) {
                    assigned_slot = preferred;
                } else {
                    for (int s = 0; s < 4; ++s) {
                        if (!room.peers[s].active) {
                            assigned_slot = s;
                            break;
                        }
                    }
                }
            }

            JoinResponsePacket resp{};
            resp.header = {0x53, 0x01, JOIN_RESPONSE};
            resp.room_id = room_id;

            if (assigned_slot == -1) {
                resp.status = 1; // FULL
                resp.peer_id = 0xFF;
                sendto(sockfd, &resp, sizeof(resp), 0, (struct sockaddr*)&cliaddr, len);
                std::cout << "[Join Denied] Room " << room_id << " is full.\n";
            } else {
                resp.status = 0; // SUCCESS
                resp.peer_id = assigned_slot;
                room.peers[assigned_slot] = { (uint8_t)assigned_slot, cliaddr, now, true };
                client_to_room[ep] = { room_id, (uint8_t)assigned_slot };

                sendto(sockfd, &resp, sizeof(resp), 0, (struct sockaddr*)&cliaddr, len);
                std::cout << "[Join Accepted] Room " << room_id << " Assigned Slot " << assigned_slot << "\n";
            }
        }
        else if (hdr->type == HEARTBEAT) {
            auto it = client_to_room.find(ep);
            if (it != client_to_room.end()) {
                auto& room = rooms[it->second.first];
                room.peers[it->second.second].last_seen = now;
                sendto(sockfd, buffer, n, 0, (struct sockaddr*)&cliaddr, len); // Echo
            }
        }
        else if (hdr->type == PLAYER_STATE || hdr->type == ACTION_EVENT || hdr->type == DAMAGE_EVENT) {
            auto it = client_to_room.find(ep);
            if (it != client_to_room.end()) {
                uint32_t room_id = it->second.first;
                uint8_t sender_slot = it->second.second;
                auto& room = rooms[room_id];
                room.peers[sender_slot].last_seen = now;

                // Fan out to all other peers in the same room
                for (int s = 0; s < 4; ++s) {
                    if (s != sender_slot && room.peers[s].active) {
                        sendto(sockfd, buffer, n, 0, (struct sockaddr*)&room.peers[s].addr, sizeof(sockaddr_in));
                    }
                }
            }
        }
        else if (hdr->type == DISCONNECT) {
            auto it = client_to_room.find(ep);
            if (it != client_to_room.end()) {
                uint32_t room_id = it->second.first;
                uint8_t slot = it->second.second;
                rooms[room_id].peers[slot].active = false;
                client_to_room.erase(it);
                std::cout << "[Client Left] Room " << room_id << " Slot " << (int)slot << "\n";
            }
        }
    }

    close(sockfd);
    return 0;
}
